src/hal/src/external_memory/mod.rs

```diff
@@ -0,0 +1,498 @@
+    type Error = &'static str;
+    fn try_from(tuple: (ExternalMemoryType, PlatformMemory)) -> Result<Self, Self::Error> {
+        match tuple {
+            #[cfg(unix)]
+            (ExternalMemoryType::OpaqueFd, PlatformMemory::Fd(fd)) => Ok(Self::OpaqueFd(fd)),
+            #[cfg(windows)]
+            (ExternalMemoryType::OpaqueWin32, PlatformMemory::Handle(handle)) => {
+                Ok(Self::OpaqueWin32(handle))
+            }
+            #[cfg(windows)]
+            (ExternalMemoryType::OpaqueWin32Kmt, PlatformMemory::Handle(handle)) => {
+                Ok(Self::OpaqueWin32Kmt(handle))
+            }
+            #[cfg(windows)]
+            (ExternalMemoryType::D3D11Texture, PlatformMemory::Handle(handle)) => {
+                Ok(Self::D3D11Texture(handle))
+            }
+            #[cfg(windows)]
+            (ExternalMemoryType::D3D11TextureKmt, PlatformMemory::Handle(handle)) => {
+                Ok(Self::D3D11TextureKmt(handle))
+            }
+            #[cfg(windows)]
+            (ExternalMemoryType::D3D12Heap, PlatformMemory::Handle(handle)) => {
+                Ok(Self::D3D12Heap(handle))
+            }
+            #[cfg(windows)]
+            (ExternalMemoryType::D3D12Resource, PlatformMemory::Handle(handle)) => {
+                Ok(Self::D3D12Resource(handle))
+            }
+            #[cfg(any(target_os = "linux", target_os = "android"))]
+            (ExternalMemoryType::DmaBuf, PlatformMemory::Fd(fd)) => Ok(Self::DmaBuf(fd)),
+            #[cfg(target_os = "android")]
+            (ExternalMemoryType::AndroidHardwareBuffer, PlatformMemory::Fd(fd)) => {
+                Ok(Self::AndroidHardwareBuffer(fd))
+            }
+            (ExternalMemoryType::HostAllocation, PlatformMemory::Ptr(ptr)) => {
+                Ok(Self::HostAllocation(ptr))
+            }
+            (ExternalMemoryType::HostMappedForeignMemory, PlatformMemory::Ptr(ptr)) => {
+                Ok(Self::HostMappedForeignMemory(ptr))
+            }
+            // Without this on non unix or windows platform, this will trigger error for unreachable pattern
+            #[allow(unreachable_patterns)]
+            _ => Err("Wrong handle type and platform memory combination"),
+        }
+    }
+}
+
+#[cfg(unix)]
+impl std::convert::TryFrom<(ExternalMemoryType, Fd)> for ExternalMemory {
+    type Error = &'static str;
+    fn try_from(tuple: (ExternalMemoryType, Fd)) -> Result<Self, Self::Error> {
+        ExternalMemory::try_from((tuple.0, PlatformMemory::from(tuple.1)))
+    }
+}
+
+#[cfg(windows)]
+impl std::convert::TryFrom<(ExternalMemoryType, Handle)> for ExternalMemory {
+    type Error = &'static str;
+    fn try_from(tuple: (ExternalMemoryType, Handle)) -> Result<Self, Self::Error> {
+        ExternalMemory::try_from((tuple.0, PlatformMemory::from(tuple.1)))
+    }
+}
+
+impl std::convert::TryFrom<(ExternalMemoryType, Ptr)> for ExternalMemory {
```

Comment:
We don't need to implement all of these. If the user has a `Handle`, let them call `PlatformMemory::from` and just use the single way.

Reply:
Yes, they are not needed. Removed.